### PR TITLE
Deprecate CUID targets in add_slack_variables transformation

### PR DESCRIPTION
**Pull request by emma58**
## Summary/Motivation:

Deprecates the requirement that targets for `core.add_slack_variables` are a list of ComponentUIDs. Instead, they are expected to be a Constraint or list of Constraints.

## Changes proposed in this PR:
- Changes to ConfigBlock for handling transformation arguments
- Deprecates cuid targets in favor of Constraints or lists of Constraints
- Updates tests accordingly

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/core/plugins/transform/add_slack_vars.py [hunk 1]
```diff
@@ -5,11 +5,40 @@
...
+from pyomo.common.deprecation import deprecation_warning
 
-from random import randint
-
-import logging
-logger = logging.getLogger('pyomo.core')
+def target_list(x):
+    deprecation_msg = ("In future releases ComponentUID targets will no longer "
+                       "be supported. Specify targets as a Constraint or list "
+                       "of Constraints.")
```
Comment (jsiirola) [inline pyomo/core/plugins/transform/add_slack_vars.py:16]: This should include the name of the transformation in the message.

File: pyomo/core/plugins/transform/add_slack_vars.py [hunk 2]
```diff
@@ -5,11 +5,40 @@
...
+    if isinstance(x, ComponentUID):
+        deprecation_warning(deprecation_msg)
```
Comment (jsiirola) [inline pyomo/core/plugins/transform/add_slack_vars.py:18]: We should probably only emit the deprecation message once for every call to the transformation.  Maybe something like:
```python
if deprecation_msg:
    deprecation_warning(deprecation_msg)
    deprecation_msg = None
```

File: pyomo/core/plugins/transform/add_slack_vars.py [hunk 3]
```diff
@@ -5,11 +5,40 @@
...
+        # [ESJ 07/15/2020] We have to just pass it through because we need the
+        # instance in order to be able to do anything about it...
+        return [ x ]
+    elif isinstance(x, (Constraint, _ConstraintData)):
+        return [ x ]
+    elif hasattr(x, '__iter__'):
+        ans = []
+        for i in x:
+            if isinstance(i, ComponentUID):
+                deprecation_warning(deprecation_msg)
```
Comment (jsiirola) [inline pyomo/core/plugins/transform/add_slack_vars.py:28]: ...same as above: only emit the deprecation message once for every call?

File: pyomo/core/tests/transform/test_add_slacks.py [hunk 4]
```diff
@@ -260,28 +260,50 @@ def test_target_objective_create_using(self):
...
+    def test_error_for_non_constraint_target_in_list(self):
+        m = self.makeModel()
+        self.assertRaisesRegexp(
+            ValueError,
+            "Expected Constraint or list of Constraints.\n\tRecieved "
+            "<class 'pyomo.core.base.var.SimpleVar'>",
+            TransformationFactory('core.add_slack_variables').apply_to,
+            m,
+            targets=[m.rule1, m.x]
+            )
```
Comment (jsiirola) [inline pyomo/core/tests/transform/test_add_slacks.py:299]: You should add a test that confirms that ComponentUID still works, and that it emits a deprecation warning (probably using LoggingIntercept)

**Comment by emma58** [comment]
@jsiirola, OK, I think I addressed everything. Good point about only yelling at people once...

**Pull request merged**